icu4c/source/i18n/units_complexconverter.cpp

```diff
@@ -196,37 +231,54 @@ MaybeStackVector<Measure> ComplexUnitsConverter::convert(double quantity,
             // Add the last element, not an integer:
             Formattable formattableQuantity(quantity * sign);
             // Measure takes ownership of the MeasureUnit*
-            MeasureUnit *type = new MeasureUnit(units_[i]->copy(status).build(status));
-            if (result.emplaceBackAndCheckErrorCode(status, formattableQuantity, type, status) ==
-                nullptr) {
+            MeasureUnit *type = new MeasureUnit((units_[i])->unitImpl->copy(status).build(status));
+            if (unordered_result.emplaceBackAndCheckErrorCode(
+                    status, units_[i]->index, Measure(formattableQuantity, type, status)) == nullptr) {
                 // Ownership wasn't taken
                 U_ASSERT(U_FAILURE(status));
                 delete type;
             }
             if (U_FAILURE(status)) {
                 return result;
             }
-            U_ASSERT(result.length() == i + 1);
-            U_ASSERT(result[i] != nullptr);
+            U_ASSERT(unordered_result.length() == i + 1);
+            U_ASSERT(unordered_result[i] != nullptr);
         }
     }
 
-    MaybeStackVector<Measure> orderedResult;
-    int32_t unitsCount = outputUnits_.length();
-    U_ASSERT(unitsCount == units_.length());
-    Measure **arr = result.getAlias();
-    // O(N^2) is fine: mixed units' unitsCount is usually 2 or 3.
-    for (int32_t i = 0; i < unitsCount; i++) {
-        for (int32_t j = i; j < unitsCount; j++) {
-            // Find the next expected unit, and swap it into place.
-            U_ASSERT(result[j] != nullptr);
-            if (result[j]->getUnit() == *outputUnits_[i]) {
-                if (j != i) {
-                    Measure *tmp = arr[j];
-                    arr[j] = arr[i];
-                    arr[i] = tmp;
-                }
-            }
+    // Sort the unordered_result
+
+    // NOTE:
+    //  This comparator is used to sort the units in ascending order according to their indices.
+    auto ascendingOrderByIndexComparator = [](const void *, const void *left, const void *right) {
+        const auto *leftPointer = static_cast<const MeasureWithIndex *const *>(left);
+        const auto *rightPointer = static_cast<const MeasureWithIndex *const *>(right);
+
+         int32_t diff = (*leftPointer)->index -  (*rightPointer)->index;
+         if (diff == 0) {
+            return 0;
+        }
+        return diff > 0 ? 1 : -1;
+    };
+
+    uprv_sortArray(unordered_result.getAlias(),     //
+                   unordered_result.length(),       //
+                   sizeof unordered_result[0],      //
+                   ascendingOrderByIndexComparator, //
+                   nullptr,                         //
+                   false,                           //
+                   &status                          //
+    );
+
+    // FIXME: unordered_result is in the same order as units_, is it not? And
```

Comment:
This sorting logic looked wrong to me, when I analysed it, and then I found there isn't a unit test for C++ yet? Add the two unit tests to check if this logic is correct?

Reply:
it is just comparing the indices ??? what is the problem about it ?

Reply:
As mentioned, I missed that you're calling uprv_sortArray again - I'm sorry. Shall I delete the fixmes, or do you want to do it?

Reply:
FIXMEs deleted: I've also removed the second call to sort.